packages/core/test/invariant.test.ts

```diff
@@ -0,0 +1,221 @@
+    actor.send({ type: 'loadUser' });
+
+    expect(spy).toHaveBeenCalledWith(new Error('User not loaded'));
+
+    expect(actor.getSnapshot().value).toEqual('idle');
+  });
+
+  it('transitions as normal if the invariant does not fail', () => {
+    const machine = createMachine({
+      initial: 'idle',
+      states: {
+        idle: {
+          on: {
+            loadUser: {
+              target: 'userLoaded',
+              actions: assign({ user: () => ({ name: 'David' }) })
+            }
+          }
+        },
+        userLoaded: {
+          invariant: (x) => {
+            if (!x.context.user) {
+              throw new Error('User not loaded');
+            }
+          }
+        }
+      }
+    });
+    const spy = jest.fn();
+
+    const actor = createActor(machine);
+    actor.subscribe({
+      error: spy
+    });
+    actor.start();
+
+    actor.send({ type: 'loadUser' });
+
+    expect(spy).not.toHaveBeenCalled();
+
+    expect(actor.getSnapshot().value).toEqual('userLoaded');
+  });
+
+  it('throws an error and does not transition if the invariant fails on a transition within the state', () => {
+    const machine = createMachine({
+      initial: 'userLoaded',
+      states: {
+        userLoaded: {
+          initial: 'active',
+          states: {
+            active: {
+              on: {
+                deactivate: 'inactive'
+              }
+            },
+            inactive: {
+              entry: assign({ user: null })
+            }
+          },
+          invariant: (x) => {
+            if (!x.context.user) {
+              throw new Error('User not loaded');
+            }
+          },
+          entry: assign({ user: { name: 'David' } })
+        }
+      }
+    });
+    const spy = jest.fn();
+
+    const actor = createActor(machine);
+    actor.subscribe({
+      error: spy
+    });
+    actor.start();
+
+    actor.send({ type: 'deactivate' });
+
+    expect(spy).toHaveBeenCalledWith(new Error('User not loaded'));
+    expect(actor.getSnapshot().value).toEqual({ userLoaded: 'active' });
+  });
+
+  it('does not throw an error when exiting a state with an invariant if the exit action clears the context', () => {
+    const machine = createMachine({
+      initial: 'userLoaded',
+      states: {
+        userLoaded: {
+          invariant: (x) => {
+            if (!x.context.user) {
+              throw new Error('User not loaded');
+            }
+          },
+          entry: assign({ user: { name: 'David' } }),
+          exit: assign({ user: null }),
+          on: {
+            logout: 'idle'
+          }
+        },
+        idle: {}
+      }
+    });
+    const spy = jest.fn();
+
+    const actor = createActor(machine);
+    actor.subscribe({
+      error: spy
+    });
+    actor.start();
+
+    actor.send({ type: 'logout' });
+
+    expect(spy).not.toHaveBeenCalled();
+    expect(actor.getSnapshot().value).toEqual('idle');
+  });
+
+  it('interacts correctly with parallel states', () => {
```

Comment:
this test title tells me nothing ;p could we use something more specific? what kind of actual granular-ish behavior is being tested here?

Reply:
[879ae52](https://github.com/statelyai/xstate/pull/4944/commits/879ae529e2f354af0b8e773d6cf4c96d78de1b50)